app-backend/tool/src/main/scala/ast/codec/MapAlgebraLeafCodecs.scala

```diff
@@ -0,0 +1,47 @@
+package com.azavea.rf.tool.ast.codec
+
+import com.azavea.rf.tool.ast._
+
+import io.circe._
+import io.circe.syntax._
+
+import java.security.InvalidParameterException
+
+trait MapAlgebraLeafCodecs {
+  implicit def mapAlgebraDecoder: Decoder[MapAlgebraAST]
+  implicit def mapAlgebraEncoder: Encoder[MapAlgebraAST]
+
+  /** TODO: Add codec paths besides `raster source` and `operation` when supported */
+  implicit def mapAlgebraLeafDecoder = Decoder.instance[MapAlgebraAST.MapAlgebraLeaf] { ma =>
+    ma._type match {
+      case Some("src") =>
+        ma.as[MapAlgebraAST.Source]
+      case Some("const") =>
+        ma.as[MapAlgebraAST.Constant]
+      case _ =>
+        throw new InvalidParameterException(s"Unrecognized AST: $ma")
```

Comment:
Since `HCursor.as` returns an `Either`, you could likewise return an `Either[DecodingError, A]` here instead of throwing.